quil-rs/src/program/mod.rs

```diff
@@ -217,36 +238,64 @@ impl Program {
         Ok(expanded_program)
     }
 
+    pub fn is_protoquil(&self) -> Result<()> {
+        self.to_instructions(true)
+            .iter()
+            .find(|i| !i.is_valid_protoquil())
+            .map_or(Ok(()), |invalid_instruction| {
+                Err(ProgramError::InvalidProtoQuilInstruction(
+                    invalid_instruction.clone(),
+                ))
+            })
+    }
+
+    pub fn is_quilt(&self) -> Result<()> {
+        self.to_instructions(true)
+            .iter()
+            .find(|i| !i.is_valid_quilt())
+            .map_or(Ok(()), |invalid_instruction| {
+                Err(ProgramError::InvalidQuiltInstruction(
+                    invalid_instruction.clone(),
+                ))
+            })
+    }
+
     pub fn to_instructions(&self, include_headers: bool) -> Vec<Instruction> {
         let mut result = vec![];
 
         if include_headers {
-            result.extend(self.memory_regions.iter().map(|(name, descriptor)| {
-                Instruction::Declaration(Declaration {
-                    name: name.clone(),
-                    size: descriptor.size.clone(),
-                    sharing: descriptor.sharing.clone(),
-                })
-            }));
-            result.extend(self.frames.to_instructions());
-            result.extend(self.waveforms.iter().map(|(name, definition)| {
-                Instruction::WaveformDefinition(WaveformDefinition {
-                    name: name.clone(),
-                    definition: definition.clone(),
-                })
-            }));
-            result.extend(self.calibrations.to_instructions());
+            result.extend(self.to_headers())
         }
 
         result.extend(self.instructions.clone());
 
         result
     }
 
+    pub fn to_headers(&self) -> Vec<Instruction> {
```

Comment:
are `headers` and `instructions` good buckets to separate things by?

Reply:
nope, and I regret that decision from long ago! Even the "headers" (which is a term I invented here) are themselves instructions. I think the ability to simplify a program removes the need to separate "headers" from the other instructions, which I likely added to just trim the tens of thousands of lines of calibrations off of a program.

tl;dr I think `to_instructions()` should take no arguments and return all of the "headers" too.